Brave ads

bridiver opened the pull request:
## Submitter Checklist:

- [ ] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [ ] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [x] Windows
  - [x] macOS
  - [x] Linux
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests`) on
  - [x] Windows
  - [x] macOS
  - [x] Linux
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [x] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone

## Test Plan:


## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

bridiver commented on patches/chrome-browser-notifications-notification_platform_bridge_mac.mm.patch:
```diff
@@ -0,0 +1,12 @@
+       initWithCloseLabel:l10n_util::GetNSString(IDS_NOTIFICATION_BUTTON_CLOSE)
+             optionsLabel:l10n_util::GetNSString(IDS_NOTIFICATION_BUTTON_MORE)
+            settingsLabel:l10n_util::GetNSString(
++            notification_type == NotificationHandler::Type::BRAVE_ADS ? IDS_VIEW_MENU_MAC :
```
the main issue was getting rid of the "Settings" button, but there was a secondary issue that put 'View' inside a `More` list in the notification

tmancey commented on components/brave_ads/browser/ads_service_impl.cc:
```diff
@@ -0,0 +1,914 @@
+#if defined(OS_MACOSX)
+  client_info->platform = ads::ClientInfoPlatformType::MACOS;
+#elif defined(OS_WIN)
+  client_info->platform = ads::ClientInfoPlatformType::WIN10;
```
Is it not possible to deter win7 and win8 as if it we need to discuss with Mandar to change spec

bridiver replied:
I wouldn't let this pass privacy review with that info so I didn't want to add it and then leave a comment on my own PR that it's not ok :)

tmancey commented on components/brave_ads/browser/ads_service_impl.cc:
```diff
@@ -0,0 +1,914 @@
+#elif defined(OS_ANDROID)
+  ssid = net::android::GetWifiSSID();
+#endif
+  // TODO: Handle non UTF8 SSIDs.
```
Is this needed for 1.0?

bridiver replied:
no, the comment was just copied from some chromium code as a reminder for when we deal with non-UTF8 everywhere

bridiver replied:
it's also not critical information, mangled SSIDs won't impact anything important

tmancey commented on components/brave_ads/browser/ads_tab_helper.cc:
```diff
@@ -0,0 +1,177 @@
+    ads_service_->ClassifyPage(distiller_result->markup_info().url(),
+                               distiller_result->distilled_content().html());
+  } else {
+    // TODO(bridiver) - fall back to web_contents()->GenerateMHTML or ignore?
```
Is this a question we need answering before 1.0?

bridiver replied:
I don't think so

bridiver commented on components/brave_ads/browser/ads_service_impl.cc:
```diff
@@ -0,0 +1,914 @@
+  ResetTimer();
+}
+
+void AdsServiceImpl::OnCreate() {
```
not sure what, if anything, I need to do here yet

tmancey replied:
👍

bridiver commented on components/brave_ads/browser/ads_service_impl.cc:
```diff
@@ -0,0 +1,914 @@
+}
+
+void ConnectionClosed() {
+  // TODO(bridiver)
```
same as above, need to figure out if I need to handle this

tmancey replied:
👍

bridiver commented on components/brave_ads/browser/ads_service_impl.cc:
```diff
@@ -0,0 +1,914 @@
+}
+
+bool AdsServiceImpl::IsAdsEnabled() const {
+  return is_enabled();
```
`is_enabled` is exposed in the public API for AdsService and using the same method name as the AdsClient would create problems for this class which implements both

tmancey replied:
👍

bridiver commented on components/brave_ads/browser/ads_service_impl.cc:
```diff
@@ -0,0 +1,914 @@
+}
+
+bool AdsServiceImpl::IsNotificationsAvailable() const {
+#if BUILDFLAG(ENABLE_NATIVE_NOTIFICATIONS)
```
every OS we build for has notifications so it seemed kind of pointless to add `#if defined(OS_MACOSX)`, etc... checks that all returned `true`

bridiver replied:
as we discussed, the only information we have is whether or not notifications are available on the platform. In particular macos doesn't appear to provide an API to determine if notifications are enabled for the app or not

bridiver commented on components/brave_ads/browser/ads_service_impl.cc:
```diff
@@ -0,0 +1,914 @@
+}
+
+bool AdsServiceImpl::IsNetworkConnectionAvailable() {
+  // TODO - implement this
```
this can be implemented in a follow-up. It was added kind of late in the process and I don't want to hold this PR up longer

darkdh replied:
you can use `content::GetNetworkConnectionTracker()->IsOffline()`
And if you want to track connection change, you can use `network::NetworkConnectionTracker::NetworkConnectionObserver:: OnConnectionChanged `

bridiver commented on components/brave_ads/browser/ads_tab_helper.cc:
```diff
@@ -0,0 +1,187 @@
+  }
+}
+
+// TODO(bridiver) - what is the android equivalent of this?
```
@SergeyZhukovsky ?

bridiver commented on components/brave_ads/browser/ads_service_impl.h:
```diff
@@ -0,0 +1,190 @@
+  void Stop();
+  void ResetTimer();
+  void CheckIdleState();
+#if !defined(OS_ANDROID)
```
@SergeyZhukovsky do you know how we should handle this in android?

bridiver commented on components/brave_ads/browser/ads_service_impl.cc:
```diff
@@ -0,0 +1,933 @@
+                   profile_->IsOffTheRecord(),
+                   base::BindOnce(&Noop));
+
+// TODO(bridiver) - what do we do here for android
```
@SergeyZhukovsky ?

bridiver commented on components/brave_ads/browser/ads_service_factory.cc:
```diff
@@ -0,0 +1,83 @@
+  registry->RegisterBooleanPref(prefs::kBraveAdsEnabled, false);
+  registry->RegisterUint64Pref(prefs::kBraveAdsPerHour, 2);
+  registry->RegisterUint64Pref(prefs::kBraveAdsPerDay, 6);
+  registry->RegisterIntegerPref(prefs::kBraveAdsIdleThreshold, 20);
```
fixed

tmancey replied:
👍

darkdh commented on components/brave_ads/browser/BUILD.gn:
```diff
@@ -0,0 +1,55 @@
+      "//components/history/core/common",
+      "//components/wifi",
+      "//net",
+      "//ui/base",
```
don't we need to add `//sql` into deps?

tmancey commented on components/brave_ads/browser/ads_tab_helper.cc:
```diff
@@ -0,0 +1,187 @@
+}
+
+void AdsTabHelper::DidAttachInterstitialPage() {
+  // TODO maybe this shouldn't fire? Are we tracking times?
```
We are not tracking times

tmancey commented:
@bridiver ChangeLocale is not called when the OS locale changes

diracdeltas commented on DEPS:
```diff
@@ -18,6 +18,10 @@ deps = {
+  "vendor/bat-native-ads": "https://github.com/brave-intl/bat-native-ads.git@635cd13137b87407c97a0a825c78d941c99228cb",
+  "vendor/bat-native-usermodel": "https://github.com/brave-intl/bat-native-usermodel.git@3eb3883b0d30cbe4d6f1d907d6c76f28728eb17c",
+  "vendor/bat-native-confirmations": "https://github.com/brave-intl/bat-native-confirmations.git@9410bd8306aacf11cadddde4bb17e9a79cf29d4e",
+  "vendor/challenge-bypass-ristretto-ffi/src": "https://github.com/brave-intl/challenge-bypass-ristretto-ffi@6f33c811ef682a9c23b0134dcd2aa8b478c55231",
```
@riastradh-brave please do a quick sec review of these deps. thanks

evq replied:
I think review for `bat-native-confirmations` and `challenge-bypass-ristretto-ffi` can / should be decoupled from this PR. they are not hooked up - maybe we should remove them from the deps in this PR?

diracdeltas replied:
agree, plz remove if these aren't used and don't require review yet

tmancey replied:
Discussed with @bridiver and he is going to remove, thanks

diracdeltas commented on components/brave_ads/browser/ads_service_impl.cc:
```diff
@@ -0,0 +1,933 @@
+
+void AdsServiceImpl::ResetTimer() {
+  idle_poll_timer_.Stop();
+  idle_poll_timer_.Start(FROM_HERE,
```
sorry, this was meant for `SetTimer` not `ResetTimer`

diracdeltas replied:
i remember pointing this out in a previous PR so i think there is already a method defined for doing something every T interval with a randomized delay each time

diracdeltas replied:
to clarify, this only applies to cases where the thing-which-is-recurring involves a network request. ex: fetching ad catalog, reporting confirmations. it does not apply to local-only operations like picking an ad to show out of the local catalog or running a local classifier.

tmancey replied:
> i remember pointing this out in a previous PR so i think there is already a method defined for doing something every T interval with a randomized delay each time

Could you please point me to the method (or if you know the the randomized delay range) and I will implement the same logic on the native-ads for fetching the catalog

diracdeltas replied:
@tmancey i was thinking of https://github.com/brave/brave-core/commit/0649d184e33c070625c2a537d70a95ec662100ef, which is not reusable and also is not really correct (see my comment on the commit)

basically you want the equivalent of this (in JS):

```
function doThingWithRandomDelay(interval, maxRandomOffset, callback) {
  return setInterval(setTimeout(callback, maxRandomOffset), interval)
}
```

where `maxRandomOffset` is at least interval divided by expected-number-of-requests-per-interval across all clients.

for instance, say there are 100,000 daily active users which make one of these requests per 24 hours. 24 hours divided by 100,000 is 0.864s. then 2s is a reasonable value for maxRandomOffset.

to see why this is necessary, consider the case where there are only two users of Brave Ads, Alice and Bob:
* Alice makes her first ads catalog request at 6:00:00 and repeats it exactly every 24hours
* Bob makes his first ads catalog request at 18:00:00 and repeats it exactly every 24hours
* Brave can easily distinguish the two users based on when the request is occurring

granted this is not a big deal for ads catalog since that request shouldn't include any sensitive info, but it is important for ads confirmation (in which we want requests from the same user to be as unlinkable as possible)

tmancey replied:
Ads confirmation is not added yet and will start to be added next week. This does make sense and based upon above I now need to find out the expected number of requests per interval across all clients. Is this something we can address once ads confirmation is added?

diracdeltas replied:
@tmancey sgtm; please open a follow up issue if one is needed

tmancey replied:
@diracdeltas Right now we check for and download a new catalog every x seconds, determined by a PING value returned from the server. If for any reason a check fails we then on desktop retry after 1 minute (or mobile 2 minutes) and if that fails we we multiply the retry by 2 until we are successful (or the app is restarted where we revert to the PING value). Could you please let me know the random -/+ ve to be to cater for the above, thanks

diracdeltas replied:
@tmancey what is the range of x?

tmancey replied:
x can be any number and is passed back in the JSON of the catalog, i.e. 7200000 which in seconds equates to 7200 (2 hours) right now. Thanks

diracdeltas replied:
@tmancey `x/10` seems reasonable to me

tmancey replied:
👍 making the change now, and will update once complete

tmancey replied:
@diracdeltas fix pushed and code is as follows:

```void AdsServe::UpdateNextCatalogCheck() {
  next_retry_start_timer_in_ = 0;

  auto ping = bundle_->GetCatalogPing();
  auto rand_delay = helper::Math::Random(ping / 10);

  // Add randomized delay so that the Ad server can't correlate users by timing
  next_catalog_check_ = ping + rand_delay;

  ads_->StartCollectingActivity(next_catalog_check_);
}
```

diracdeltas commented:
[Security] does this use any data from incognito sessions for ads classifying or other purposes? if so is any of it logged?

tmancey commented:
> [Security] does this use any data from incognito sessions for ads classifying or other purposes? if so is any of it logged?
> 
> subquestions:
> 
> * are any metrics tracked for private windows? (looks like not)
> * if user clicks notification and the current window is a private window, does the landing page open in the private window?
> * if so, is the landing page time tracked in the private window?
> * is any data from private browsing written to disk?

If a tab is private browsing then no metrics are tracked, please search is_incognito to see the relevant code.

If a user clicks a notification a new tab should open, however as we do not track private tabs I do not believe this to be an issue

No data from private browsing is written to disk

bridiver commented:
> [Security] does this use any data from incognito sessions for ads classifying or other purposes? if so is any of it logged?
> 
> subquestions:
> 
> * are any metrics tracked for private windows? (looks like not)
> * if user clicks notification and the current window is a private window, does the landing page open in the private window?
> * if so, is the landing page time tracked in the private window?

https://github.com/brave/brave-core/pull/765/files#diff-3639f92629488e9f547d8ff7ead4a5f7R26
which also means that notifications will never happen in private windows

diracdeltas commented:
for any data which is written to disk that relates to browsing activity (ex: pages visited, user classifications, which ads were shown/seen, etc), does this get deleted when browsing history is cleared? 

if not we will almost certainly get external reports that this is a bug (ex: from hackerone), so please open a follow up issue for that.

tmancey commented:
> for any data which is written to disk that relates to browsing activity (ex: pages visited, user classifications, which ads were shown/seen, etc), does this get deleted when browsing history is cleared?
> 
> if not we will almost certainly get external reports that this is a bug (ex: from hackerone), so please open a follow up issue for that.

When browsing history is deleted, the Ads state is also deleted

diracdeltas commented:
Rebased this on master (2ad0900a7f4e5459921523ac0c7eade6017f0a47) and built on MacOS. Getting SEGV_MAPERR whenever Brave is closed, i can DM you the stack trace if you want.

after re-opening brave and clicking on 'restore pages', the browser immediately crashes and throws:

```
Received signal 11 SEGV_MAPERR 000000000008                          [542/1835]
0   libbase.dylib                       0x000000012855a5ee base::debug::StackTr
ace::StackTrace(unsigned long) + 174
1   libbase.dylib                       0x000000012855a6ad base::debug::StackT$
ace::StackTrace(unsigned long) + 29
2   libbase.dylib                       0x000000012812bd2a base::debug::StackT$
ace::StackTrace() + 26
3   libbase.dylib                       0x000000012855a3e1 base::debug::(anony$
ous namespace)::StackDumpSignalHandler(int, __siginfo*, void*) + 1393
4   libsystem_platform.dylib            0x00007fff5dc91f5a _sigtramp + 26
5   libbase.dylib                       0x00000001280c96ab base::AutoLock::~Au$
oLock() + 43
6   libchrome_dll.dylib                 0x000000010f38fcef brave_ads::AdsServi$
eImpl::TabClosed(SessionID) + 127
7   libchrome_dll.dylib                 0x000000010f38858f brave_ads::AdsTabHe$
per::WebContentsDestroyed() + 79
8   libcontent.dylib                    0x0000000131ed464c content::WebContent$
Impl::~WebContentsImpl() + 5132
9   libcontent.dylib                    0x0000000131ed6445 content::WebContent$
Impl::~WebContentsImpl() + 21
10  libcontent.dylib                    0x0000000131ed6569 content::WebContents
Impl::~WebContentsImpl() + 25
11  libchrome_dll.dylib                 0x00000001165ae0c5 TabStripModel::SendD
etachWebContentsNotifications(TabStripModel::DetachNotifications*) + 4261
12  libchrome_dll.dylib                 0x00000001165bc0e9 TabStripModel::Close
WebContentses(base::span<content::WebContents* const, 18446744073709551615ul>,
unsigned int) + 4249
13  libchrome_dll.dylib                 0x00000001165b1482 TabStripModel::Inter
nalCloseTabs(base::span<content::WebContents* const, 18446744073709551615ul>, u
nsigned int) + 658
14  libchrome_dll.dylib                 0x00000001165b178e TabStripModel::Close
WebContentsAt(int, unsigned int) + 286
15  libchrome_dll.dylib                 0x0000000116492667 chrome::CloseWebCont
ents(Browser*, content::WebContents*, bool) + 247
16  libchrome_dll.dylib                 0x0000000111a10f50 SessionRestoreImpl::
ProcessSessionWindows(std::__1::vector<std::__1::unique_ptr<sessions::SessionWi
ndow, std::__1::default_delete<sessions::SessionWindow> >, std::__1::allocator<
std::__1::unique_ptr<sessions::SessionWindow, std::__1::default_delete<sessions
::SessionWindow> > > >*, SessionID, std::__1::vector<SessionRestoreDelegate::Re
storedTab, std::__1::allocator<SessionRestoreDelegate::RestoredTab> >*) + 4144
...
```

bridiver commented:
rebased on master? I just rebased this yesterday so shouldn't require any other changes. I'll check to see if something happened on my rebase, but in order to build and run correctly you'll have to rebase the brave-browser branch too

bridiver commented:
I could only repro once out of about 10 tries, but I see the issue so fixing now

bridiver commented:
just did another rebase, shouldn't be any changes

diracdeltas commented:
not crashing for me anymore

bridiver commented on components/brave_ads/browser/ads_service.h:
```diff
@@ -0,0 +1,42 @@
+  AdsService() = default;
+
+  virtual bool is_enabled() const = 0;
+  virtual uint64_t ads_per_hour() const = 0;
```
these methods will be removed in a follow-up PR to use the prefs

bridiver commented:
0.59.x https://github.com/brave/brave-core/commit/88b781c0c3e179b857a8d7d1155562c13607eb4e
